File: guests/python/reorder.py
"""The reorder scene (tools/scenes/reorder.steps). THE ROOT IS A ROW so the
For's container is the only column, and column#0 names it everywhere."""

import sys
from dataclasses import dataclass

import kaya


@dataclass
class Item:
    title: str


app = kaya.App()


def on_rotate():
    first = items.keys()[0]
    items.move_to_end(first)


def on_lift():
    items.move_to_front(items.keys()[-1])


with app.window():
    items = kaya.collection(Item)
    with kaya.row():
        kaya.button("rotate", on_click=on_rotate)
        kaya.button("lift", on_click=on_lift)
        for item in items:
            kaya.label(bind=item.title)
    for key in ["a", "b", "c"]:
        items.insert(key, Item(title=key))

sys.exit(app.run())

File: guests/js/reorder.ts
// The reorder scene: THE ROOT IS A ROW, so column#0 names one widget.
//     KAYA_SELFTEST=reorder node guests/js/reorder.ts

import * as kaya from "kaya-gui";

const Item = kaya.record({ title: String }, "Item");

const app = new kaya.App();

function onRotate(): void {
  const first = items.keys()[0]!;
  items.moveToEnd(first);
}

function onLift(): void {
  items.moveToFront(items.keys().at(-1)!);
}

let items!: kaya.Collection<kaya.Fields<typeof Item.schema>, kaya.Row<typeof Item.schema>>;

app.window(() => {
  items = kaya.collection(Item);
  kaya.row(() => {
    kaya.button("rotate", { onClick: onRotate });
    kaya.button("lift", { onClick: onLift });
    for (const item of items) {
      kaya.label({ bind: item.title });
    }
  });
  for (const key of ["a", "b", "c"]) {
    items.insert(key, Item({ title: key }));
  }
});

app.run();
